server/src/main/java/org/opensearch/search/query/QueryPhase.java

```diff
@@ -440,7 +442,29 @@ protected boolean searchWithCollector(
             boolean hasFilterCollector,
             boolean hasTimeout
         ) throws IOException {
-            return QueryPhase.searchWithCollector(searchContext, searcher, query, collectors, hasFilterCollector, hasTimeout);
+            // create the top docs collector last when the other collectors are known
+            final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(searchContext, hasFilterCollector);
+            return searchWithCollector(searchContext, searcher, query, collectors, topDocsFactory, hasFilterCollector, hasTimeout);
+        }
```

Comment:
Just for my understanding -- if I want to override the `TopDocsCollectorContext` at the top of the collector tree then with this change the way to do that is provide by own `QueryPhaseSearcher` that overrides this `searchWithCollector` method, right?

Reply:
Yes, this is what we're going to do for hybrid query with [custom collector](https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/search/query/HybridQueryPhaseSearcher.java) in neural-search plugin. After this PR is merged I'll create one more PR where this change will be used, will ping you in that new PR.